Mathlib/Analysis/SpecialFunctions/Log/PosLog.lean

```diff
@@ -0,0 +1,165 @@
+    simp only [this, and_true, ge_iff_le]
+    linarith
+
+/-!
+## Estimates for Products
+-/
+/-- Estimate for `log⁺` of a product. See `Real.posLog_prod` for a variant involving
+multiple factors. -/
+theorem posLog_mul {a b : ℝ} :
+    log⁺ (a * b) ≤ log⁺ a + log⁺ b := by
+  by_cases ha : a = 0
+  · simp [ha, posLog]
+  by_cases hb : b = 0
+  · simp [hb, posLog]
+  unfold posLog
+  nth_rw 1 [← add_zero 0, log_mul ha hb]
+  exact max_add_add_le_max_add_max
+
+/-- Estimate for `log⁺` of a product. Special case of `Real.posLog_mul` where one of
+the factors is a natural number. -/
+theorem posLog_nat_mul {n : ℕ} {a : ℝ} :
+    log⁺ (n * a) ≤ log n + log⁺ a := by
+  rw [← log_of_nat_eq_posLog]
+  exact posLog_mul
+
+/-- Estimate for `log⁺` of a product. See `Real.posLog_mul` for a variant with
+only two factors. -/
+theorem posLog_prod {α : Type*} (s : Finset α) (f : α → ℝ) :
+    log⁺ (∏ t ∈ s, f t) ≤ ∑ t ∈ s, log⁺ (f t) := by
+  classical
+  apply Finset.induction (p := fun (S : Finset α) ↦ (log⁺ (∏ t ∈ S, f t) ≤ ∑ t ∈ S, log⁺ (f t)))
+  · -- Empty set
+    simp [posLog]
+  · -- Non empty set
+    intro a s ha hs
+    calc log⁺ (∏ t ∈ insert a s, f t)
+    _ = log⁺ (f a * ∏ t ∈ s, f t) := by rw [Finset.prod_insert ha]
+    _ ≤ log⁺ (f a) + log⁺ (∏ t ∈ s, f t) := posLog_mul
+    _ ≤ log⁺ (f a) + ∑ t ∈ s, log⁺ (f t) := add_le_add (by rfl) hs
+    _ = ∑ t ∈ insert a s, log⁺ (f t) := by rw [Finset.sum_insert ha]
+
+/-!
+## Estimates for Sums
+-/
+/-- Estimate for `log⁺` of a sum. See `Real.posLog_add` for a variant involving
+just two summands. -/
+theorem posLog_sum {α : Type*} (s : Finset α) (f : α → ℝ) :
+    log⁺ (∑ t ∈ s, f t) ≤ log (s.card) + ∑ t ∈ s, log⁺ (f t) := by
+  -- Trivial case: empty sum
+  by_cases hs : s = ∅
+  · simp [hs, posLog]
+  -- Nontrivial case: Obtain maximal element…
+  obtain ⟨t_max, ht_max⟩ := s.exists_max_image (fun t ↦ |f t|)
+    (Finset.nonempty_iff_ne_empty.mpr hs)
+  -- …then calculate
+  calc log⁺ (∑ t ∈ s, f t)
+  _ = log⁺ |∑ t ∈ s, f t| := by
+    rw [Real.posLog_abs]
+  _ ≤ log⁺ (∑ t ∈ s, |f t|) := by
+    apply monotoneOn_posLog (by simp) (by simp [Finset.sum_nonneg])
+    simp [Finset.abs_sum_le_sum_abs]
+  _ ≤ log⁺ (∑ t ∈ s, |f t_max|) := by
+    apply monotoneOn_posLog (by simp [Finset.sum_nonneg]) (by simp [Finset.sum_nonneg]; positivity)
+    apply Finset.sum_le_sum (fun i ih ↦ ht_max.2 i ih)
+  _ = log⁺ (s.card * |f t_max|) := by
+    simp [Finset.sum_const]
+  _ ≤ log s.card + log⁺ |f t_max| := posLog_nat_mul
+  _ ≤ log s.card + ∑ t ∈ s, log⁺ (f t) := by
+    gcongr
+    rw [posLog_abs]
+    apply Finset.single_le_sum (fun _ _ ↦ posLog_nonneg) ht_max.1
+
+/-- Estimate for `log⁺` of a sum. See `Real.posLog_sum` for a variant involving
+multiple summands. -/
+theorem posLog_add {a b : ℝ} : log⁺ (a + b) ≤ log 2 + log⁺ a + log⁺ b := by
+  convert posLog_sum (Finset.univ : Finset (Fin 2))
+    (fun i => match i with | 0 => a | 1 => b : Fin 2 → ℝ) using 1
+  <;> simp [add_assoc]
```

Comment:
```suggestion
  convert posLog_sum Finset.univ ![a, b] using 1 <;> simp [add_assoc]
```
We have a special handy notation for functions defined on `Fin 2` or `Fin 3` or whatever: `![a, b]` is the function mapping `0` to `a` and `1` to `b`.